zerver/decorator.py

```diff
@@ -795,7 +795,7 @@ def rate_limit_user(request: HttpRequest, user: UserProfile, domain: str) -> Non
     request._ratelimit_remaining = calls_remaining
     request._ratelimit_secs_to_freedom = time_reset
 
-def rate_limit(domain: str='all') -> Callable[[ViewFuncT], ViewFuncT]:
+def rate_limit_view(domain: str='all') -> Callable[[ViewFuncT], ViewFuncT]:
```

Comment:
For naming this, is your thought that we'll use this same decorator for logged-out views, or a different one?  I think this name is good if we're going to make it support unauthed views, which seems right, but just wanted to check.

Reply:
There is a TODO in the function to add non-authed limiting and it seems to make sense to do that while implementing IP-based limiting to logins (we can have the login limiting as a special case with its own limits to general IP limiting for non-authed views that would go here).

Another approach I'm thinking of is whether it wouldn't be a better design rework things a bit and get rid of the decorator, moving as much logic as possible into RateLimitMiddleware? As things stand, it's non-obvious what gets limited and what doesn't - some stuff is exluded from limiting in rest.py, and the views that get limited do so because of being wrapped in wrappers which happen to add the rate_limit wrapped inside their body. Instead, in rest.py, where we do the dispatching, could set request._rate_limit = True/False explicitly and then RateLimitMiddleware would capture the requests with this set to True and do all the work.

Reply:
After some more investigating, I'd like to alter my suggestion somewhat - don't explicityl set rate limit True/False in rest.py - RateLimitMiddleware should assume it should rate limit unless specified otherwise. So, in rest.py where we do the exclusion by setting skip_rate_limiting=True in the decorators, we'll instead set request._skip_rate_limiting=True and RateLimitMiddleware will check that. This approach makes it clear that everything is rate limited unless explicitly excluded, compared to the current approach where one needs to dig into internals of decorators to see rate_limit is applied somewhere inside.

And some more involved exclusion rules or custom rate limiting rules can be applied inside RateLimitMiddleware logic if necessary - but the point stands that we have a central place to find what the rate limiter does instead of it being scattered across decorators.

Reply:
In this approach, we may need to have 2 components to RateLimitMiddleware - to sit in 2 various places in the middleware sequence through which requests pass - RateLimitMiddlewareByIP which would sit where the current RateLimitMiddleware is, and RateLimitMiddlewareAuthed which would be just below the authentication(with 2FA) middlewares, where the user-based rate limiting would occur. It should be fairly easy to, in RateLimitMiddlewareByIP, look at the request to see if it's going to a password-checking endpoint (login/password change) and apply custom limits suitable for these endpoints.

Reply:
Made an ugly flowchart to illustrate this idea perhaps more clearly than in my lengthy descriptions above:
![Untitled Diagram](https://user-images.githubusercontent.com/45007152/55074042-8bf2c780-508f-11e9-861e-4cca882442e6.png)

Reply:
Scratch the above, for some reason I thought the middleware layers are executed after rest.py, but it's the opposite of that actually. I'll keep thinking if the middleware idea can be modified in a sensible way despite this complication.